File: bin/pickrefids.pl
#!/usr/bin/perl

#############################################
#
# Program: Pick reference genomes based on
#          metaphyler output
#
# Author: Bo Liu, Victoria Cepeda, Mathieu Almeida
#
# Mon Aug 22 13:57:01 EDT 2016
#
#############################################

use strict;
use warnings;
use FindBin qw($Bin);


#----------------------------------------#
# read command line options
#----------------------------------------#
my $blast = "";
my $coverage = 3;
my $readlen = 100;
if (scalar @ARGV == 3) {
    ($blast, $coverage, $readlen) = @ARGV;
} else {
    Usage();
}

#----------------------------------------#


#----------------------------------------#
# information of reference genomes
#----------------------------------------#
my %seq2tid = ();    # sequence id to taxonomy id
my %tid2seqs = ();   # taxonomy id to sequences
my %tid2name = ();   # tax id to name
my %tid2sp = ();     # tax id to species id
open(FH, "$Bin/../refseq/tid2par.tab") or die("Could not open $Bin/tid2par.tab\n");
foreach my $line (<FH>) {
    chomp $line;
    my ($seq, $tid, $sp, $ge, $name) = split("\t", $line);
    $seq2tid{$seq} = $tid;
    $tid2seqs{$tid}{$seq} = 1;
    $tid2name{$tid} = $name;
    if ($sp eq "NA") { $sp = $tid;}
    $tid2sp{$tid} = $sp;
}
close FH;
#----------------------------------------#


#----------------------------------------#
# abundance for each reference genome
#----------------------------------------#
my %tid2num = ();   # # of reads classified within each tax id

# read tab delimited file
open(FH, "$blast") or die("Could not open $blast.\n");
foreach my $line (<FH>) {
    chomp $line;
    my ($qid, $rid, $pct, $hspl) = split("\t", $line);
    if ($pct < 95 && $hspl < 35) { next;}
    $rid =~ /^(\S+)\_\S+\_\S+$/;
    
    $rid = $1;
    my $tid = $seq2tid{$rid};
    $tid2num{$tid}++;
}
close FH;
#----------------------------------------#


#----------------------------------------#
# based on the abundance, pick reference
#----------------------------------------#
my $total = 0;
my %sps = ();
foreach my $tid (sort {$tid2num{$b} <=> $tid2num{$a}} keys %tid2num) {

    my $num = $tid2num{$tid};
    my $marker_cumulated_size = 10000.0;
    if (( ($num*$readlen) / $marker_cumulated_size) < $coverage) { last;} # skip if prediceted coverage is below threshold
    if ($total > 10000) { last;} # skip if more than 10000 genomes selected
    
    my $sp = $tid2sp{$tid};
    if (exists $sps{$sp}) {
	if ($sps{$sp} >= 5) { next;}  # skip if more than 5 genomes from same species selected
    }
    
    $sps{$sp}++;
    $total++;

    #put the same counts for the chromosome and all molecules related (like plasmids) 
    
    foreach my $seq (keys %{$tid2seqs{$tid}}) {
   	  print "$seq\t$tid\t$num\t$tid2name{$tid}\n";
    }
}



exit;


sub Usage {
    die("
Usage:
       perl pickrefids.pl <blast file> <partition>

Options:
       <blast>        Query sequences in FASTA format to be classified.
       <partition>    Suppose you used a quarter of total data to estimate the
                      taxonomic profile. Then partition number is 4.

Output:
       A list of NCBI genome IDs, which will be used as reference genomes.

");
}

# todo
# pick reference genomes based on the estimation of depth of coverage
# allow users to tune this parameter, right now it's hard coded

File: bin/pickrefids.py
#!/usr/bin/env python

#############################################
#
# Program: Pick reference genomes based on
#          metaphyler output
#
# Author: Mathieu Almeida, Bo Liu, Victoria Cepeda
#
# Thu Sep 28 07:55:00 EDT 2017
#
#############################################

import sys
import os
import operator

#============================
#calculate median coverage
#============================
def median(L):
   if sum(L)>0:
         
         srt = sorted(L)
         n = len(L)
         mid = n//2
         
         if n % 2:
            return float(srt[mid])
         else:
            med = (srt[mid] + srt[mid-1]) / 2.0 
            return float(med)
   
   else:
      return 0.0

#============================
#calculte average coverage
#============================
def average(x):
   sum_x = sum(l)
   len_x = len(l)
   return( sum_x / float(len_x))
 
#============================
#extract metaphyler taxa data
#============================
def extract_reference_data():
   seq2tid={}
   tid2seqs={}
   tid2name={}
   tid2sp={}
   tid2seq={}
   pathbin="/".join(os.path.realpath(__file__).split('/')[:-2])   

   with open(pathbin + "/refseq/tid2par.tab", "rt") as reffile:
     for line in reffile:
        
        seqid, taxonid, species, genus, name = line.split("\t")
        
        seq2tid[seqid] = taxonid.strip()
        tid2name[taxonid] = name.strip()
        tid2sp[taxonid] = species.strip()
        if not tid2seq.get(taxonid):
           tid2seq[taxonid] = []
        tid2seq[taxonid].append(seqid)
                   
   return seq2tid, tid2name, tid2sp, tid2seq


#========================================
#initalize marker coverage arrays
#========================================
def init_marker_data(marker2coverage,markertotalcount):
   #add blast list marker2coverage
   pathbin="/".join(os.path.realpath(__file__).split('/')[:-2])
   #marker2coverage={}
   #markertotalcount={}
   with open(pathbin + "/src/metaphyler/markers/markers.refseq.dna", "rt") as markerfile:
      header=""
      seq=""
      flag=0
      for line in markerfile:
         if line[0]==">":
           #first line
           if header!="" and flag == 0:
                  ref = "_".join(header.split('_')[0:2]
                  print "%s" % [0]
                  marker2coverage[header] = [0] * (len(seq)-1)
              #markertotalcount[header] = 0
           header= line.strip()[1:]
           if marker2coverage.has_key(header):
               flag=0
           else:
               flag=1
               header=""
           seq=""
         else:
            if flag ==0:
                seq+= line.strip()
      #last line
      if flag ==0:
         ref = "_".join(header.split('_')[0:2])
         marker2coverage[header] = [0] * (len(seq)-1)            
      #markertotalcount[header]=0
   return marker2coverage


#===========================================================
#extract blast information from the read to marker alignment
#===========================================================
def extract_marker_data(blast_filename, marker2coverage, markertotalcount ):
  
   with open(blast_filename, "rt") as blastfile:
      for line in blastfile:
         items = line.split("\t")
         marker = items[1]
         mstart = int(items[8])
         mend = int(items[9])
         if int(mstart) < int(mend):
            pos=int(mstart)
            while pos < int(mend):
               if !marker2coverage.has_key(marker):
             
               marker2coverage[marker][pos-1] += 1
               #markertotalcount[marker]=0
               pos += 1
         else:
            pos=int(mend)
            while pos < int(mstart):
               marker2coverage[marker][pos-1] += 1
               #markertotalcount[marker]=0
               pos += 1
         markertotalcount[marker] += 1
         #markertotalcount[marker]=0
   #return marker2coverage, markertotalcount
   

#============================================================
#generate reference genome list data for metaphyler
#============================================================
def create_refid_file(marker2coverage, markertotalcount, seq2tid, tid2name, tid2sp, tid2seq, covthreshold, nbsamespeciesthreshold):
   total_ref_cov={}   
   #list_species=[]
   list_ref_median_cov={}

   for marker in marker2coverage:
      ref = "_".join(marker.split('_')[:2])
      if not total_ref_cov.get(ref):
         total_ref_cov[ref] = marker2coverage[marker]
      else:
         total_ref_cov[ref] = total_ref_cov.get(ref) + marker2coverage[marker]

   #calculate median
   for ref in total_ref_cov:
      list_ref_median_cov[ref] = float(median(total_ref_cov[ref]))

   #write result
   for ref, median_cov in sorted(list_ref_median_cov.items(), key=operator.itemgetter(1), reverse=True):
      taxa = seq2tid[ref]
      name = tid2name[taxa]
      #species = tid2sp[taxa]
      
      if median_cov >= 0.0:#covthreshold:
         #species_count = list_species.count(species) 
         #if species_count < nbsamespeciesthreshold:
            print("%s\t%s\t%.1f\t%s" % (ref,taxa, median_cov,name))
            #print("%s\t%.1f\t%s" % (ref,median_cov,name))
            #print ("%s\t%.1f" % (ref,median_cov))
            #for refid in tid2seq[taxa]:
            #   print("%s\t%.1f\t%s" % (refid,median_cov,name))
               #print("%s\t%s\t%.1f\t%s" % (refid,taxa, median_cov,name)) 
            #list_species.append(species)

#===========================
#MAIN
#===========================
def main():
   """Main program function
   """
   blast_filename = sys.argv[1]
   covthreshold = float(sys.argv[2])
   nbsamespeciesthreshold=5
   marker2coverage={}
   markertotalcount={}
   seq2tid, tid2name, tid2sp,tid2seq = extract_reference_data()
   extract_marker_data(blast_filename, marker2coverage, markertotalcount )
   init_marker_data(marker2coverage)
   create_refid_file(marker2coverage, markertotalcount, seq2tid, tid2name, tid2sp, tid2seq, covthreshold, nbsamespeciesthreshold)
          

if __name__ == '__main__':
    main()
 
